devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/dependency/ApplicationDeploymentClasspathBuilder.java

```diff
@@ -88,70 +99,164 @@ public static void initConfigurations(Project project) {
+                        final DefaultDependencyArtifact dep = new DefaultDependencyArtifact();
+                        dep.setExtension("properties");
+                        dep.setType("properties");
+                        dep.setName(name);
+
+                        final DefaultExternalModuleDependency gradleDep = new DefaultExternalModuleDependency(
+                            group, name, d.getTarget().getVersion(), null);
+                        gradleDep.addArtifact(dep);
+
+                        for (ResolvedArtifact a : project.getConfigurations().detachedConfiguration(gradleDep)
+                            .getResolvedConfiguration().getResolvedArtifacts()) {
+                            if (a.getName().equals(name)) {
+                                try {
+                                    platformImports.addPlatformProperties(group, name, null, "properties", d.getTarget().getVersion(),
+                                        a.getFile().toPath());
+                                } catch (AppModelResolverException e) {
+                                    throw new GradleException("Failed to import platform properties " + a.getFile(), e);
+                                }
+                                break;
+                            }
+                        }
+                    }
+                });
+            });
         }
-        return runtimeConfig;
     }
 
-    public Configuration getDeploymentConfiguration() {
-        String deploymentConfigurationName = ToolingUtils.toDeploymentConfigurationName(runtimeConfig.getName());
-        Configuration deploymentConfig = project.getConfigurations().findByName(deploymentConfigurationName);
-        if (deploymentConfig != null) {
-            return deploymentConfig;
+    private void setUpRuntimeConfiguration() {
+        if (project.getConfigurations().findByName(this.runtimeConfigurationName) == null) {
+            project.getConfigurations().register(this.runtimeConfigurationName, configuration ->
+                configuration.extendsFrom(
+                    project.getConfigurations()
+                        .getByName(ApplicationDeploymentClasspathBuilder.getBaseRuntimeConfigName(mode))));
         }
+    }
 
-        final Collection<ExtensionDependency> allExtensions = getConditionalDependenciesEnabler().getAllExtensions();
-        Set<ExtensionDependency> extensions = collectFirstMetQuarkusExtensions(getRuntimeConfiguration(), allExtensions);
-        // Add conditional extensions
-        for (ExtensionDependency knownExtension : allExtensions) {
-            if (knownExtension.isConditional()) {
-                extensions.add(knownExtension);
-            }
-        }
+    private void setUpDeploymentConfiguration() {
+        if (project.getConfigurations().findByName(this.deploymentConfigurationName) == null) {
+            project.getConfigurations().register(this.deploymentConfigurationName, configuration -> {
```

Comment:
Would `maybeCreate` not work?

Reply:
TBH, I'm not sure why these get registered instead of created anyway.

Reply:
> `#maybeCreate` already handles potential existing one. That's the "maybe" part.

This code path is hit multiple times but if this code is allowed to run multiple times it will create duplicate sets of metadata.

> TBH, I'm not sure why these get registered instead of created anyway.

register and create do the same thing, but create is eager whereas register is deferred. Deferring doesn't buy us much yet as Quarkus digs into these Configurations almost immediately right now.

Reply:
> This code path is hit multiple times but if this code is allowed to run multiple times it will create duplicate sets of metadata.

Which is why you generally create them in one place up front

> Deferring doesn't buy us much yet as Quarkus digs into these Configurations almost immediately right now.

It also has an important distinction in using them for wiring things up though.

Say you want to use these in a task, - `QuarkusDev#quarkusDevConfiguration` e.g.  To leverage deferred configuration, we'd need to either:

1. Change that field from `Configuration` to `Provider<Configuration>`
2. Use create rather than register and keep the field as `Configuration`

Using this delayed creation gains nothing here I believe, and is less flexible than just creating it.

Reply:
> > This code path is hit multiple times but if this code is allowed to run multiple times it will create duplicate sets of metadata.
> 
> Which is why you generally create them in one place up front

I believe this would require a complete rewrite of ApplicationDeploymentClasspathBuilder and how it gets used by the tasks. My mind started to wander in this direction and that's when I put my foot on the brakes.

> > Deferring doesn't buy us much yet as Quarkus digs into these Configurations almost immediately right now.
> 
> It also has an important distinction in using them for wiring things up though.
> 
> Say you want to use these in a task, - `QuarkusDev#quarkusDevConfiguration` e.g. To leverage deferred configuration, we'd need to either:
> 
>     1. Change that field from `Configuration` to `Provider<Configuration>`
> 
>     2. Use create rather than register and keep the field as `Configuration`
> 
> 
> Using this delayed creation gains nothing here I believe, and is less flexible than just creating it.

The configuration is deferred internally in the ConfigurationContainer. As soon as they're requested by name, they're created and the registered action is run. This means that the Configuraton is as immediately available as it would be with "create".

I can change "register" to "create" and it would functionally remain the same.

Reply:
> I can change "register" to "create" and it would functionally remain the same.

Never said it would.  But as-is, this code is creating deferred objects (proxies really) for something that does not have to be a proxy.  2 objects is always worse when 1 will suffice.  

Now if this deferred creation actually gained something that would be a different story.  But it does not here. 

> I believe this would require a complete rewrite of ApplicationDeploymentClasspathBuilder and how it gets used by the tasks.

No idea what you are thinking of here.  Specifics?  As part of the constructor injection work I started, I actually moved the creation (ok, ok.. registration) of Configurations way earlier.  And it worked utterly fine.  I have this:

```
public class QuarkusPlugin implements Plugin<Project> {
    ...

    @Override
    public void apply(Project project) {
        ...

        createConfigurations(project);
        registerTasks(project, quarkusExt);
    }
}
```

Reply:
> Now if this deferred creation actually gained something that would be a different story. But it does not here.

That I can see anyway.  Please correct me if I missed something

Reply:
> > Now if this deferred creation actually gained something that would be a different story. But it does not here.
> 
> That I can see anyway. Please correct me if I missed something

You haven't missed anything. The point was to create an obvious point for another refactor, but since that aesthetic has become a problem rather than a help, I'll simply roll it back to "create".

Reply:
> No idea what you are thinking of here. Specifics? As part of the constructor injection work I started, I actually moved the creation (ok, ok.. registration) of Configurations way earlier. And it worked utterly fine. I have this:

I was trying to limit the size and scope of the MR. As it stands, the bulk of the logic is not changed, and simply moved to a new location that makes it obvious how it can be further improved. In order for me to break the creation away from the fetching of these Configurations, I'd have to make design choices, which could further mire the PR in debate.

I can very clearly move these "setUp*Configuration" function bodies to another spot, but then PlatformImports will need some thought. They're tightly coupled and created per project, which is its own issue as well, and requires a that the metadata potentially being moved somewhere better. Or, possibly, the logic that creates PlatformImports should be rewritten to run later, or using different metadata.

Reply:
I don't think I'm holding this up.  At least I hope not.  I don't think it should.

You asked a question, I answered and we had a discussion ;)   I totally agree its best to do incremental, targeted pull requests.

Reply:
Sorry, I misunderstood your intention. I thought you were asking for changes. I was being defensive because the result of this conversation, if I tried to address your questions with code, would have resulted in a lot of work. At that point I'd probably turn the PR into a draft, or close it and start another one, to go over ideas.

I have changed "register" to "create" as that was easy enough to do.